src/miner.erl

```diff
@@ -41,16 +38,52 @@
     terminate/2
 ]).
 
+
+-type metadata_v1() ::
+    {integer(), blockchain_block:hash()}.
+
+-type metadata_v2() ::
+    #{
+        timestamp      => integer(),
+        seen           => binary(),
+        bba_completion => binary(),
+        head_hash      => blockchain_block:hash(),
+        snapshot_hash  => binary()
+     }.
+
+-type metadata() ::
+    [{J :: pos_integer(), M :: metadata_v2() | metadata_v1()}].
+
+-type swarm_keys() ::
+    {libp2p_crypto:pubkey(), libp2p_crypto:sig_fun()}.
+
+-type create_block_error() ::
+      stale_hash
+    | multiple_hashes.
+
+-type create_block_ok() ::
+    #{
+        address               =>  libp2p_crypto:pubkey_bin(),
+        unsigned_binary_block =>  binary(),
+        signature             =>  binary(),
+        pending_txns          =>  blockchain_txn:txns(),
+        invalid_txns          =>  blockchain_txn:txns()
+    }.
+
+-type create_block_result() ::
+      {ok, create_block_ok()}
+    | {error, create_block_error()}.
+
 -record(state, {
-    %% NOTE: a miner may or may not participate in consensus
+    % NOTE: a miner may or may not participate in consensus
```

Comment:
please don't do this, it messes up auto-formatting.  % is for line end comments %% is for comments with matching indentation